akka-cluster-sharding/src/main/resources/reference.conf

```diff
@@ -99,12 +99,36 @@ akka.cluster.sharding {
     # the region with least shards must be greater than (>) the `rebalanceThreshold`
     # for the rebalance to occur.
     # 1 gives the best distribution and therefore typically the best choice.
-    # Increasing the threshold can result in quicker rebalance but has the
-    # drawback of increased difference between number of shards (and therefore load)
-    # on different nodes before rebalance will occur.
     rebalance-threshold = 1
 
-    # The number of ongoing rebalancing processes is limited to this number.
+    # The maximum number of shards to start rebalancing on each rebalance decision.
+    # Increasing this will increase the speed at which rebalancing occurs.
+    # This differs from max-simultaneous-rebalance, in that the latter applies across
+    # multiple rebalance intervals, whereas this only applies to a single rebalance
+    # interval decision.
+    # The smaller of this number and the limit decided by the rebalance-factor will
+    # be used. If this is zero, then just the rebalance-factor will be used. If both
+    # this and the rebalance factor are zero, then the rebalance-threshold will be
+    # used as the limit.
+    max-rebalance-per-interval = 0
```

Comment:
To make it easier to understand I think we should think less about compatibility and more about what would be a better behavior. For Akka 2.6.0 we can change default configuration values and behavior. We will document the change in the migration guide, and possibly describe how to configure the previous behavior.

What if we let `rebalance-threshold` only be about if if a rebalance should occur or not (diff > threshold)?

If someone want same as before they can set `max-rebalance-per-interval` to same value as the threshold?

What would be good defaults in general, assuming 1000 shards over 100 nodes, and also 1000 shards over 10 nodes?
```
rebalance-threshold = 1
max-rebalance-per-interval = 0
rebalance-factor = 0.1
max-simultaneous-rebalance = 10
```

For 1000 shards over 100 nodes, adding one new node => 1 shard per iteration
For 1000 shards over 10 nodes, adding one new node => 10 shards per iteration

Reply:
I have no idea what good defaults will be. I agree that not supporting the old behaviour would make things both easier to understand, and it will make the code easier to understand too.

Another one to consider is 1000 shards over 20 nodes. On the first 9 iterations, you'll have 5 shards rebalanced. At that point, (assuming the rebalanced nodes have been reallocated) you'll have 10 nodes with 45 shards, and 10 nodes with 50 shards. So, it's then going to move 2 shards from each of the ones with 50 shards to the nodes with 45 shards over the next 10 iterations, and then it'll be balanced. But the result is that it's moved 20 more shards than needed to be moved (optimal is to move 47, but in total we moved 67).

I think, in general, the least optimal allocations, in terms of work performed, if the rebalance factor is greater than the ratio of nodes to shards, then you'll get an inefficient rebalancing. So that could be used to help decide what the ideal number should be.

If we are going to change the behaviour, then maybe it might be a good idea to implement behaviour along the lines of what @chbatey was suggesting.

Currently, we limit the number of shards rebalanced by taking the midway point between the smallest and largest nodes. Perhaps, rather than that, we should limit it by the ideal final outcome - so, the maximum number of shards to move will be limited to reducing the size of the largest node to the total number of shards divided by the total number of nodes. This will ensure we never do more work than needed. Also, currently the rebalance factor is applied to the number of nodes allocated to the shard. This means as time progresses, the number of shards rebalanced reduces on each iteration. Perhaps it would be better to instead, apply the factor to the ideal number of shards per node, that way it will be constant throughout the rebalance.

Reply:
I agree that it should be possible to predict a better if looking at the full picture. It should be pretty common to join more than one node and then it's always a risk of moving shards more than necessary if only looking at largest/smallest and moving too many in each iteration. As you mentioned that should be a new allocation strategy.

I still think this has potential to be better than current implementation, if we could wrap it up by simplifying and and finding decent defaults.